overcast/runner/__init__.py

```diff
@@ -128,6 +128,99 @@ def get_creds_from_env():
 #    d['cacert'] = os.environ.get('OS_CACERT', None)
     return d
 
+
+class Node(object):
+
+    def __init__(self, name, info, keypair, userdata, runner, record_resource=None):
+        self.record_resource = lambda *args, **kwargs: None
+        self.name      = name
+        self.info      = info
+        self.runner    = runner
+        self.keypair   = keypair
+        self.userdata  = userdata
+        self.server_id = None
+        self.fip_ids   = set()
+        self.port_ids  = set()
+        self.fip_address = None
+        self.server_status = None
+        self.nc = self.runner.get_nova_client()
+        if self.info['image'] in self.runner.mappings.get('images', {}):
+            self.info['image'] = self.runner.mappings['images'][self.info['image']]
+
+        if self.info['flavor'] in self.runner.mappings.get('flavors', {}):
+            self.info['flavor'] = self.runner.mappings['flavors'][self.info['flavor']]
+        self.image = self.nc.images.get(self.info['image'])
+        self.flavor = self.nc.flavors.get(self.info['flavor'])
+
+    ##
+    # This one poll the server and return server status.
+    ##
+    def poll(self, desired_status = 'ACTIVE'):
+        if self.server_status != desired_status:
+            self.server_status = self.nc.servers.get(self.server_id).status
+        return self.server_status
+
+    ##
+    # Cleaner: This method remove server, fip, port etc.
+    #
+    # We could keep fip and may be ports (ports are getting deleted with current
+    # neutron client), but that is going to be bit more complex to make sure
+    # right port is assigned to right fip etc, so atm, just removing them.
+    ##
+    def clean(self):
+        for fip_id in self.fip_ids:
+            delete_floating_ip(fip_id)
+            self.fip_ids.discard(fip_id)
```

Comment:
Modifying a collection while iterating over it is asking for trouble. In many cases, it won't even be allowed.

You can just:
    for fip_id in self.fip_ids:
        self.runner.delete_floating_ip(fip_id)
    
    self.fip_ids = set()